modules/csg/csg_shape.cpp

```diff
@@ -32,6 +32,306 @@
 
 #include "core/math/geometry_2d.h"
 
+#include "thirdparty/manifold/src/manifold/include/manifold.h"
+#include "thirdparty/manifold/src/utilities/include/public.h"
+
+// Static helper functions.
+
+struct CSGBrush {
+	struct Face {
+		Vector3 vertices[3];
+		Vector2 uvs[3];
+		AABB aabb;
+		bool smooth = false;
+		bool invert = false;
+		int material = 0;
+	};
+
+	Vector<Face> faces;
+	Vector<Ref<Material>> materials;
+
+	manifold::Manifold manifold;
+	enum {
+		MANIFOLD_PROPERTY_POS_X = 0,
+		MANIFOLD_PROPERTY_POS_Y,
+		MANIFOLD_PROPERTY_POS_Z,
+		MANIFOLD_PROPERTY_UV_X,
+		MANIFOLD_PROPERTY_UV_Y,
+		MANIFOLD_PROPERTY_MATERIAL,
+		MANIFOLD_PROPERTY_SMOOTH,
+		MANIFOLD_PROPERTY_INVERT,
+		MANIFOLD_MAX
+	};
+	static constexpr int MANIFOLD_TRIANGLE = 3;
+
+	inline void _regen_face_aabbs() {
+		for (int i = 0; i < faces.size(); i++) {
+			faces.write[i].aabb = AABB();
+			faces.write[i].aabb.position = faces[i].vertices[0];
+			faces.write[i].aabb.expand_to(faces[i].vertices[1]);
+			faces.write[i].aabb.expand_to(faces[i].vertices[2]);
+		}
+	}
+
+	void pack_manifold(const float p_snap, HashMap<uint32_t, Ref<Material>> &r_materials) {
```

Comment:
@elalish Any recommendations to improve this? I got stuck translating your tutorial.

Reply:
Did you see my review below? Does it make sense?

Reply:
I wasn't able to understand your tutorial, it's probably something on my part.